src/libraries/System.Net.Connections/tests/System.Net.Connections.Tests/Sockets/SocketsConnectionFactoryTests.cs

```diff
@@ -0,0 +1,261 @@
+        }
+
+        // TODO: On OSX and Windows7 connection failures seem to fail with unexpected SocketErrors that are mapped to NetworkError.Unknown. This needs an investigation.
+        // Related: https://github.com/dotnet/runtime/pull/40565
+        public static bool PlatformHasReliableConnectionFailures => !PlatformDetection.IsOSX && !PlatformDetection.IsWindows7 && !PlatformDetection.IsFreeBSD;
+
+        [ConditionalFact(nameof(PlatformHasReliableConnectionFailures))]
+        public async Task ConnectAsync_WhenRefused_ThrowsNetworkException()
+        {
+            using Socket notListening = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+            int port = notListening.BindToAnonymousPort(IPAddress.Loopback);
+            var endPoint = new IPEndPoint(IPAddress.Loopback, port);
+
+            using SocketsConnectionFactory factory = new SocketsConnectionFactory(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+            NetworkException ex = await Assert.ThrowsAsync<NetworkException>(() => factory.ConnectAsync(endPoint).AsTask());
+            Assert.Equal(NetworkError.ConnectionRefused, ex.NetworkError);
+        }
+
+        [OuterLoop] // TimedOut and HostNotFound is slow on Windows
+        [ConditionalFact(nameof(PlatformHasReliableConnectionFailures))]
+        public async Task ConnectAsync_WhenHostNotFound_ThrowsNetworkException()
+        {
+            using SocketsConnectionFactory factory = new SocketsConnectionFactory(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+            // Unassigned as per https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt
+            int unusedPort = 8;
+            DnsEndPoint endPoint = new DnsEndPoint(System.Net.Test.Common.Configuration.Sockets.InvalidHost, unusedPort);
+
+            NetworkException ex = await Assert.ThrowsAsync<NetworkException>(() => factory.ConnectAsync(endPoint).AsTask());
+            Assert.Equal(NetworkError.HostNotFound, ex.NetworkError);
+        }
+
+        [OuterLoop] // TimedOut and HostNotFound is slow on Windows
+        [ConditionalFact(nameof(PlatformHasReliableConnectionFailures))]
+        public async Task ConnectAsync_TimedOut_ThrowsNetworkException()
+        {
+            using SocketsConnectionFactory factory = new SocketsConnectionFactory(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+            IPEndPoint doesNotExist = new IPEndPoint(IPAddress.Parse("1.2.3.4"), 23);
+
+            // SocketError.TimedOut currently maps to SocketError.Unknown, so no asserion
+            await Assert.ThrowsAsync<NetworkException>(() => factory.ConnectAsync(doesNotExist).AsTask());
+        }
+
+        // On Windows, connection timeout takes 21 seconds. Abusing this behavior to test the cancellation logic
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)] 
```

Comment:
Does this test not work on other platforms? Seems like it should.

Reply:
On Linux connection "timeout" comes immediately for the non-existing IP address `1.2.3.4`. It won't even work with the following code:

```C#
CancellationTokenSource cts = new CancellationTokenSource();
OperationCanceledException ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(async () => {
        await factory.ConnectAsync(doesNotExist, cancellationToken: cts.Token);
        cts.Cancel();
    });
```

Reply:
Ok, we should add a comment explaining this